apis/oam/v1alpha2/core_types.go

```diff
@@ -0,0 +1,375 @@
+type Component struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ComponentSpec   `json:"spec,omitempty"`
+	Status ComponentStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ComponentList contains a list of Component.
+type ComponentList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Component `json:"items"`
+}
+
+// A ComponentParameterValue specifies a value for a named parameter. The
+// associated component must publish a parameter with this name.
+type ComponentParameterValue struct {
+	// Name of the component parameter to set.
+	Name string `json:"name"`
+
+	// Value to set.
+	Value intstr.IntOrString `json:"value"`
+}
+
+// A ComponentTrait specifies a trait that should be applied to a component.
+type ComponentTrait struct {
+	// A Trait that will be created for the component
+	// +kubebuilder:validation:EmbeddedResource
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Trait runtime.RawExtension `json:"trait"`
+}
+
+// A ScopeReference refers to an OAM scope resource.
+type ScopeReference struct {
+	// APIVersion of the referenced scope.
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of the referenced scope.
+	Kind string `json:"kind"`
+
+	// Name of the referenced scope.
+	Name string `json:"name"`
+
+	// UID of the referenced scope.
+	// +optional
+	UID *types.UID `json:"uid,omitempty"`
+}
+
+// A ComponentScope specifies a scope in which a component should exist.
+type ComponentScope struct {
+	// A ScopeReference must refer to an OAM scope resource.
+	ScopeReference ScopeReference `json:"scopeRef"`
+}
+
+// An ApplicationConfigurationComponent specifies a component of an
+// ApplicationConfiguration. Each component is used to instantiate a workload.
+type ApplicationConfigurationComponent struct {
+	// ComponentName specifies a component of which an ApplicationConfiguration
+	// should consist. The named component must exist.
+	ComponentName string `json:"componentName"`
+
+	// ParameterValues specify values for the the specified component's
+	// parameters. Any parameter required by the component must be specified.
+	// +optional
+	ParameterValues []ComponentParameterValue `json:"parameterValues,omitempty"`
+
+	// Traits of the specified component.
+	// +optional
+	Traits []ComponentTrait `json:"traits,omitempty"`
+
+	// Scopes in which the specified component should exist.
+	// +optional
+	Scopes []ComponentScope `json:"scopes,omitempty"`
+}
+
+// An ApplicationConfigurationSpec defines the desired state of a
+// ApplicationConfiguration.
+type ApplicationConfigurationSpec struct {
+	// Components of which this ApplicationConfiguration consists. Each
+	// component will be used to instantiate a workload.
+	Components []ApplicationConfigurationComponent `json:"components"`
+}
+
+// A WorkloadReference refers to an OAM workload resource.
+type WorkloadReference struct {
```

Comment:
do we need to define 3 identical references struct? There is another identical ResourceReference struct in the  core_workload_types.go

Reply:
I recall being on the fence about this. I think I went with three identical types because the docstring here ("A WorkloadReference refers to an OAM workload resource.") is used to generate CRD and API type documentation (e.g. https://doc.crds.dev/github.com/crossplane/crossplane and https://crossplane.io/docs/master/api.html).

On reflection, the docstring for the actual field that uses this type will be included too, so we can probably use a single type here.